code/game/objects/items/weapons/RCD.dm

```diff
@@ -382,3 +388,278 @@ RCD
 	name = "admin RCD"
 	max_matter = INFINITY
 	matter = INFINITY
+
+
+// Ranged RCD
+
+
+/obj/item/weapon/rcd/arcd
+	name = "advanced rapid-construction-device (ARCD)"
+	desc = "A prototype RCD with ranged capability and extended capacity"
+	max_matter = 300
+	matter = 300
+	delay_mod = 0.6
+	ranged = TRUE
+	icon_state = "arcd"
+
+
+/obj/item/weapon/rcd/arcd/afterattack(atom/A, mob/user, proximity)
+	if(!(A in view(7, get_turf(user))))
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
+		return 0
+	if((isturf(A) && A.density && mode==RCD_DECONSTRUCT) || (isturf(A) && !A.density) || (istype(A, /obj/machinery/door/airlock) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/grille) || (istype(A, /obj/structure/window) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/girder))
+		user.Beam(A,icon_state="rped_upgrade",time=30)
```

Comment:
If this giant IF is really needed at least wrap it in some is_valid_target function, but i'm pretty sure there's base rcd function that already does it.